hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestPoolBasedHierarchicalResourceLockManager.java

```diff
@@ -0,0 +1,577 @@
+          fail("Write lock acquisition failed: " + e.getMessage());
+        }
+      }, executor);
+
+      // Wait for write lock to be acquired
+      writeLockAcquired.await();
+
+      // Second thread tries to acquire read lock
+      CompletableFuture<Void> future2 = CompletableFuture.runAsync(() -> {
+        try {
+          // This should block until write lock is released
+          readLockBlocked.set(true);
+          try (HierarchicalResourceLock lock = lockManager.acquireReadLock(FlatResource.SNAPSHOT_GC_LOCK, key)) {
+            readLockAcquired.countDown();
+          }
+        } catch (Exception e) {
+          fail("Read lock acquisition failed: " + e.getMessage());
+        }
+      }, executor);
+
+      // Wait for both threads to complete
+      future1.get(5, TimeUnit.SECONDS);
+      future2.get(5, TimeUnit.SECONDS);
+
+      assertTrue(readLockBlocked.get());
+      assertEquals(0, readLockAcquired.getCount());
+
+    } finally {
+      executor.shutdown();
+    }
+  }
+
+  /**
+   * Test lock state after closing.
+   */
+  @Test
+  public void testLockStateAfterClose() throws Exception {
+    String key = "test-key-6";
+
+    HierarchicalResourceLock lock = lockManager.acquireReadLock(FlatResource.SNAPSHOT_DB_LOCK, key);
+    assertTrue(lock.isLockAcquired());
+
+    lock.close();
+    assertFalse(lock.isLockAcquired());
+  }
+
+  /**
+   * Test double close doesn't cause issues.
+   */
+  @Test
+  public void testDoubleClose() throws Exception {
+    String key = "test-key-7";
+
+    HierarchicalResourceLock lock = lockManager.acquireWriteLock(FlatResource.SNAPSHOT_GC_LOCK, key);
+    assertTrue(lock.isLockAcquired());
+
+    // First close
+    lock.close();
+    assertFalse(lock.isLockAcquired());
+
+    // Second close should not throw exception
+    lock.close();
+    assertFalse(lock.isLockAcquired());
+  }
+
+  /**
+   * Test different resource types can be locked independently.
+   */
+  @ParameterizedTest
+  @EnumSource(FlatResource.class)
+  public void testDifferentResourceTypes(FlatResource resource) throws Exception {
+    String key = "test-key-" + resource.name();
+
+    try (HierarchicalResourceLock lock = lockManager.acquireWriteLock(resource, key)) {
+      assertNotNull(lock);
+      assertTrue(lock.isLockAcquired());
+    }
+  }
+
+
+  /**
+   * Test different keys on same resource type can be locked concurrently.
+   */
+  @Test
+  public void testDifferentKeysOnSameResource() throws Exception {
+    String key1 = "test-key-8a";
+    String key2 = "test-key-8b";
+
+    try (HierarchicalResourceLock lock1 = lockManager.acquireWriteLock(FlatResource.SNAPSHOT_GC_LOCK, key1);
+         HierarchicalResourceLock lock2 = lockManager.acquireWriteLock(FlatResource.SNAPSHOT_GC_LOCK, key2)) {
+
+      assertNotNull(lock1);
+      assertNotNull(lock2);
+      assertTrue(lock1.isLockAcquired());
+      assertTrue(lock2.isLockAcquired());
+    }
+  }
+
+  /**
+   * Test configuration parameters are respected.
```

Comment:
the test implementation itself does not ensure soft limit and hard limit are respected, only that a HierarchicalResourceLock is acquired without exceptions.